src/bootloaders/shim-systemd.c

```diff
@@ -0,0 +1,313 @@
+{
+        (void)manager;
+        if (!exists_identical(shim_dst_host, shim_src))
+                return true;
+        if (!exists_identical(systemd_dst_host, systemd_src))
+                return true;
+        return false;
+}
+
+static bool make_layout(const BootManager *manager)
+{
+        char *boot_root = boot_manager_get_boot_dir((BootManager *)manager);
+        char path[PATH_MAX];
+        snprintf(path, PATH_MAX, "%s%s", boot_root, DST_DIR);
+        if (!nc_mkdir_p(path, 00755)) {
+                goto fail;
+        }
+        snprintf(path, PATH_MAX, "%s%s", boot_root, KERNEL_DST_DIR);
+        if (!nc_mkdir_p(path, 00755)) {
+                goto fail;
+        }
+        snprintf(path, PATH_MAX, "%s%s", boot_root, SYSTEMD_ENTRIES);
+        if (!nc_mkdir_p(path, 00755)) {
+                goto fail;
+        }
+        return true;
+fail:
+        LOG_FATAL("Failed to make dir: %s", path);
+        return false;
+}
+
+static bool shim_systemd_install(const BootManager *manager)
+{
+        char varname[9];
+
+        if (!make_layout(manager)) {
+                LOG_FATAL("Cannot create layout");
+                return false;
+        }
+
+        if (!copy_file_atomic(shim_src, shim_dst_host, 00644)) {
+                LOG_FATAL("Cannot copy %s to %s", shim_src, shim_dst_host);
+                return false;
+        }
+        if (!copy_file_atomic(systemd_src, systemd_dst_host, 00644)) {
+                LOG_FATAL("Cannot copy %s to %s", systemd_src, systemd_dst_host);
+                return false;
+        }
+
+        if (bootvar_create(BOOT_DIRECTORY, shim_dst_esp, varname, 9)) {
+                LOG_FATAL("Cannot create EFI variable");
+                return false;
+        }
+
+        return true;
+}
+
+static bool shim_systemd_update(const BootManager *manager)
+{
+        return shim_systemd_install(manager);
+}
+
+static bool shim_systemd_remove(const BootManager *manager)
+{
+        (void)manager;
+        fprintf(stderr, "%s is not implemented\n", __func__);
+        return true;
+}
+
+static bool shim_systemd_init(const BootManager *manager)
+{
+        size_t len;
+        char *prefix, *boot_root;
+
+        if (bootvar_init())
+                return false;
+
+        /* init systemd-class since we're reusing it for kernel install.
+         * specific values do not matter as long as sd_class is not used to
+         * install the bootloaders themselves. */
+        static BootLoaderConfig systemd_config = {.vendor_dir = "systemd",
+                                                  .efi_dir = "/usr/lib/systemd/boot/efi",
+                                                  .efi_blob = "systemd-boot" EFI_SUFFIX,
+                                                  .name = "systemd-boot" };
+        sd_class_init(manager, &systemd_config);
+
+        prefix = strdup(boot_manager_get_prefix((BootManager *)manager));
+        len = strlen(prefix);
+        if (len > 0 && prefix[len - 1] == '/')
```

Comment:
Shouldn't you be using `basename` or `parent` APIs from Nica instead of mangling the strings? :)